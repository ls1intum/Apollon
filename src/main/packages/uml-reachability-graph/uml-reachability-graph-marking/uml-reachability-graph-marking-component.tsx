import React, { FunctionComponent } from 'react';
import { UMLReachabilityGraphMarking } from './uml-reachability-graph-marking';
import { Multiline } from '../../../utils/svg/multiline';
import { ThemedRect, ThemedPath, ThemedPolyline } from '../../../components/theme/themedComponents';

export const UMLReachabilityGraphMarkingComponent: FunctionComponent<Props> = ({ element, scale, fillColor }) => (
  <g>
    <ThemedRect
      rx={10 * scale}
      ry={10 * scale}
      width="100%"
      height="100%"
      strokeColor={element.strokeColor}
      fillColor={fillColor || element.fillColor}
    />
    <Multiline
      x={element.bounds.width / 2}
      y={element.bounds.height / 2}
      width={element.bounds.width}
      height={element.bounds.height}
      fontWeight="bold"
      fill={element.textColor}
      lineHeight={16 * scale}
      capHeight={11 * scale}
    >
      {element.name}
    </Multiline>
    {element.isInitialMarking && (
      <g>
        <marker
          id={`marker-${element.id}`}
          viewBox={`0 0 ${30 * scale} ${30 * scale}`}
          markerWidth={22 * scale}
          markerHeight={30 * scale}
          refX={30 * scale}
          refY={15 * scale}
          orient="auto"
          markerUnits="strokeWidth"
        >
          <ThemedPath
            d={`M0,${29 * scale} L${30 * scale},${15 * scale} L0,${1 * scale}`}
            fillColor="none"
            strokeColor={element.strokeColor}
          />
        </marker>
        <ThemedPolyline
          points={`-${50 * scale},-${50 * scale} ${3 * scale},${3 * scale}`}
          strokeColor={element.strokeColor}
          fillColor="none"
          strokeWidth={1 * scale}
          markerEnd={`url(#marker-${element.id})`}
        />
      </g>
    )}
  </g>
);

interface Props {
  element: UMLReachabilityGraphMarking;
  scale: number;
  fillColor?: string;
}
