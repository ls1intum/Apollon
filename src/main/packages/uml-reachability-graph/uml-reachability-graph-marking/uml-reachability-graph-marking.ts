import { ILayer } from '../../../services/layouter/layer';
import { ILayoutable } from '../../../services/layouter/layoutable';
import { UMLElement } from '../../../services/uml-element/uml-element';
import { UMLElementType } from '../../uml-element-type';
import * as Apollon from '../../../typings';
import { ReachabilityGraphElementType } from '..';
import { DeepPartial } from 'redux';
import { calculateNameBounds } from '../../../utils/name-bounds';

export class UMLReachabilityGraphMarking extends UMLElement {
  type: UMLElementType = ReachabilityGraphElementType.ReachabilityGraphMarking;
  isInitialMarking: boolean;

  constructor(values?: DeepPartial<UMLReachabilityGraphMarking>) {
    super(values);
    this.isInitialMarking = values?.isInitialMarking || false;
  }

  serialize(children?: UMLElement[]): Apollon.UMLReachabilityGraphMarking {
    return {
      ...super.serialize(),
      type: this.type as keyof typeof ReachabilityGraphElementType,
      isInitialMarking: this.isInitialMarking,
    };
  }

  deserialize<T extends Apollon.UMLModelElement>(values: T, children?: Apollon.UMLModelElement[]) {
    const assert = (v: Apollon.UMLModelElement): v is Apollon.UMLReachabilityGraphMarking =>
      v.type === ReachabilityGraphElementType.ReachabilityGraphMarking;
    if (!assert(values)) {
      return;
    }

    super.deserialize(values, children);
    this.isInitialMarking = values.isInitialMarking;
  }

  render(canvas: ILayer): ILayoutable[] {
    this.bounds = calculateNameBounds(this, canvas);
    return [this];
  }
}
