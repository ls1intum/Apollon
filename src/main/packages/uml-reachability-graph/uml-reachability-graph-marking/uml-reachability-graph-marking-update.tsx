import React, { Component, ComponentClass } from 'react';
import { connect } from 'react-redux';
import { compose } from 'redux';
import { Button } from '../../../components/controls/button/button';
import { TrashIcon } from '../../../components/controls/icon/trash';
import { Textfield } from '../../../components/controls/textfield/textfield';
import { I18nContext } from '../../../components/i18n/i18n-context';
import { localized } from '../../../components/i18n/localized';
import { ModelState } from '../../../components/store/model-state';
import { styled } from '../../../components/theme/styles';
import { UMLElementRepository } from '../../../services/uml-element/uml-element-repository';
import { UMLReachabilityGraphMarking } from './uml-reachability-graph-marking';
import { Divider } from '../../../components/controls/divider/divider';
import { ColorButton } from '../../../components/controls/color-button/color-button';
import { StylePane } from '../../../components/style-pane/style-pane';

interface OwnProps {
  element: UMLReachabilityGraphMarking;
}

type StateProps = {};

interface DispatchProps {
  update: typeof UMLElementRepository.update;
  delete: typeof UMLElementRepository.delete;
}

type Props = OwnProps & StateProps & DispatchProps & I18nContext;

const enhance = compose<ComponentClass<OwnProps>>(
  localized,
  connect<StateProps, DispatchProps, OwnProps, ModelState>(null, {
    update: UMLElementRepository.update,
    delete: UMLElementRepository.delete,
  }),
);

const Flex = styled.div`
  display: flex;
  align-items: baseline;
  justify-content: space-between;
`;

type State = { colorOpen: boolean };

class UmlReachabilityGraphMarkingUpdate extends Component<Props, State> {
  state = { colorOpen: false };

  private toggleColor = () => {
    this.setState((state) => ({
      colorOpen: !state.colorOpen,
    }));
  };

  render() {
    const { element } = this.props;

    return (
      <div>
        <section>
          <Flex>
            <Textfield value={element.name} onChange={this.rename(element.id)} autoFocus />
            <ColorButton onClick={this.toggleColor} />
            <Button color="link" tabIndex={-1} onClick={this.delete(element.id)}>
              <TrashIcon />
            </Button>
          </Flex>
          <Divider />
        </section>
        <StylePane
          open={this.state.colorOpen}
          element={element}
          onColorChange={this.props.update}
          lineColor
          textColor
          fillColor
        />
        <section>
          <label htmlFor="toggleIsInitialMarking">
            <input
              id="toggleIsInitialMarking"
              type="checkbox"
              checked={element.isInitialMarking}
              onChange={this.toggleIsInitialMarking(element.id)}
            />
            {this.props.translate('packages.ReachabilityGraph.ReachabilityGraphIsInitialMarking')}
          </label>
        </section>
      </div>
    );
  }

  private rename = (id: string) => (value: string) => {
    this.props.update(id, { name: value });
  };

  private toggleIsInitialMarking = (id: string) => (event: React.FormEvent<HTMLInputElement>) => {
    this.props.update<UMLReachabilityGraphMarking>(id, { isInitialMarking: event.currentTarget.checked });
  };

  private delete = (id: string) => () => {
    this.props.delete(id);
  };
}

export const UMLReachabilityGraphMarkingUpdate = enhance(UmlReachabilityGraphMarkingUpdate);
