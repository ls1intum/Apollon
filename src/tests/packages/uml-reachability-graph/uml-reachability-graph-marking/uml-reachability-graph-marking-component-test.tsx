import React from 'react';
import { wrappedRender } from '../../../test-utils/render';
import { UMLReachabilityGraphMarking } from '../../../../main/packages/uml-reachability-graph/uml-reachability-graph-marking/uml-reachability-graph-marking';
import { UMLReachabilityGraphMarkingComponent } from '../../../../main/packages/uml-reachability-graph/uml-reachability-graph-marking/uml-reachability-graph-marking-component';

describe('uml-reachability-graph-marking-component', () => {
  let element: UMLReachabilityGraphMarking;
  beforeEach(() => {
    element = new UMLReachabilityGraphMarking({});
  });

  it('render the uml-reachability-graph-marking-component', () => {
    const { baseElement } = wrappedRender(
      <svg>
        <UMLReachabilityGraphMarkingComponent element={element} scale={1.0} />
      </svg>,
    );
    expect(baseElement).toMatchSnapshot();
  });

  it('render with isInitialMarking false', () => {
    element.isInitialMarking = false;
    const { container } = wrappedRender(
      <svg>
        <UMLReachabilityGraphMarkingComponent element={element} scale={1.0} />
      </svg>,
    );
    expect(container.querySelectorAll('polyline')).toHaveLength(0);
  });

  it('render with isInitialMarking true', () => {
    element.isInitialMarking = true;
    const { container } = wrappedRender(
      <svg>
        <UMLReachabilityGraphMarkingComponent element={element} scale={1.0} />
      </svg>,
    );
    expect(container.querySelectorAll('polyline')).toHaveLength(1);
  });
});
