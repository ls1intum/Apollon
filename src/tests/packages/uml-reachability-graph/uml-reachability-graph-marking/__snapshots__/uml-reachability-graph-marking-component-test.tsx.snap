// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`uml-reachability-graph-marking-component render the uml-reachability-graph-marking-component 1`] = `
<body>
  <div>
    <svg>
      <g>
        <rect
          height="100%"
          rx="10"
          ry="10"
          stroke="black"
          width="100%"
        />
        <text
          font-weight="bold"
          height="100"
          pointer-events="none"
          text-anchor="middle"
          width="200"
          x="100"
          y="50"
        >
          <tspan>
            Â 
          </tspan>
        </text>
      </g>
    </svg>
  </div>
</body>
`;
